app/src/main/java/org/mozilla/reference/browser/tabs/SyncedTabsFragment.kt

```diff
@@ -0,0 +1,271 @@
+        scope.launch {
+            accountManager.syncNowAsync(SyncReason.User).await()
+        }
+    }
+
+    fun updateUI() {
+        val accountManager = requireContext().components.backgroundServices.accountManager
+        val account = accountManager.authenticatedAccount()
+
+        if (account != null && !accountManager.accountNeedsReauth()) {
+            // Signed-in, no problems.
+            if (SyncEnginesStorage(requireContext()).getStatus()[SyncEngine.Tabs] == true) {
+                // Tabs engine enabled.
+                showTabs()
+            } else {
+                // Tab engine disabled
+                showInfoMessage("Please enable tab syncing.")
+            }
+        } else if (account != null && accountManager.accountNeedsReauth()) {
+            // Signed-in, need to re-authenticate.
+            showInfoMessage("Please re-authenticate.")
+        } else {
+            // Signed-out.
+            showInfoMessage("Connect with a Firefox Account.")
+        }
+    }
+
+    private fun showInfoMessage(msg: String) {
+        synced_tabs_status.text = msg
+        synced_tabs_list.visibility = View.GONE
+        synced_tabs_status.visibility = View.VISIBLE
+    }
+
+    private fun showTabs() {
+        synced_tabs_list.visibility = View.VISIBLE
+        synced_tabs_status.visibility = View.GONE
+        scope.launch {
+            val syncedTabs = syncedTabsFeature.getSyncedTabs()
+            scope.launch(Dispatchers.Main) {
+                if (syncedTabs.isEmpty() &&
+                    accountManager.authenticatedAccount()?.deviceConstellation()?.state()?.otherDevices?.isEmpty()
+                    == true) {
+                    showInfoMessage("Connect another device.")
+                } else {
+                    val adapter = SyncedTabsAdapter(
+                        this@SyncedTabsFragment,
+                        syncedTabs
+                    )
+                    recyclerView.adapter = adapter
+                }
+            }
+        }
+    }
+
+    private class SyncedTabsAccountObserver(private val inner: SyncedTabsFragment) : AccountObserver {
+        override fun onLoggedOut() {
+            inner.updateUI()
+        }
+
+        override fun onAuthenticated(account: OAuthAccount, authType: AuthType) {
+            inner.updateUI()
+        }
+
+        override fun onAuthenticationProblems() {
+            inner.updateUI()
+        }
+    }
+
+    private class SyncedTabsSyncObserver(private val inner: SyncedTabsFragment) : SyncStatusObserver {
+        override fun onIdle() {
+            // A Sync just finished, update our adapter.
+            inner.showTabs()
+        }
+
+        override fun onError(error: Exception?) {
+            // Do nothing.
+        }
+
+        override fun onStarted() {
+            // Do nothing.
+        }
+    }
+
+    private class SyncedTabsAdapter(
+        private val clickListener: View.OnClickListener,
```

Comment:
We shouldn't do this because the child holds on to the reference of the fragment and can make the fragment live longer than we want.

We should follow the presenter/interactor pattern instead. There are a couple examples around r-b but also a-c.